src/Raven.Server/Documents/Indexes/Persistence/ConverterBase.cs

```diff
@@ -219,9 +108,120 @@ ValueType GetValueTypeUnlikely(Type type)
                 return valueType;
             }
 
-            return GetValueTypeUnlikely(type);
+            valueType = GetValueTypeUnlikely(value);
+            _valueTypeCache.Put(type, valueType);
+            return valueType;
         }
 
+        internal static ValueType GetValueTypeUnlikely(object value)
```

Comment:
This should never be called from the outside. It was an inner function specifically for that purpose.